### PR TITLE
EntityCoordinates

### PR DESCRIPTION
**Pull request by gradientvera**
Supersedes #1273
Polishes `EntityCoordinates` a bit, adds some useful methods to it, unit tests and makes Transform have `EntityCoordinates`
I'd like to get this merged as soon as possible so we can start moving stuff over.

### CODE REVIEW

File: Robust.Shared/GameObjects/Components/Transform/TransformComponent.cs [hunk 1]
```diff
@@ -281,6 +281,46 @@ public Vector2 WorldPosition
             }
         }
 
+        [ViewVariables(VVAccess.ReadWrite)]
+        public EntityCoordinates Coordinates
+        {
+            get => new EntityCoordinates(_parent.IsValid() ? _parent : Owner.Uid, LocalPosition);
```
Comment (PJB3005) [inline Robust.Shared/GameObjects/Components/Transform/TransformComponent.cs]: This is not correct. If we have to return a set of coordinates for entities without parents then it should be either:

1. Our own entity with 0 offset
2. A null entity.

File: Robust.Shared/GameObjects/Components/Transform/TransformComponent.cs [hunk 2]
```diff
@@ -281,6 +281,46 @@ public Vector2 WorldPosition
...
+                else
+                {
+                    var worldCoords = value.ToMapPos(_entityManager);
+
+                    var newPos = Parent!.InvWorldMatrix.Transform(worldCoords);
+
+                    // float rounding error guard, if the offset is less than 1mm ignore it
+                    if ((newPos - _localPosition).LengthSquared < 10.0E-3)
+                        return;
+                }
```
Comment (PJB3005) [inline Robust.Shared/GameObjects/Components/Transform/TransformComponent.cs]: This entire `else` block:
1. isn't even doing anything
2. doesn't have a purpose now that entity coordinates correctly bakes in an exact position.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: Robust.Shared/GameObjects/Components/Transform/TransformComponent.cs (hunk 1 above)
Comment (gradientvera) [reply Robust.Shared/GameObjects/Components/Transform/TransformComponent.cs]: 1 is what I was going for with this, I thought this was the correct way to go about it. What should I do?

File: Robust.Shared/GameObjects/Components/Transform/TransformComponent.cs (hunk 2 above)
Comment (gradientvera) [reply Robust.Shared/GameObjects/Components/Transform/TransformComponent.cs]: Removed now

File: Robust.Shared/GameObjects/Components/Transform/TransformComponent.cs (hunk 1 above)
Comment (gradientvera) [reply Robust.Shared/GameObjects/Components/Transform/TransformComponent.cs]: Nevermind, I figured out what you meant. Made it so it returns its own entity with 0 offset now

**Pull request merged**